contracts/contracts/strategies/balancer/BaseBalancerStrategy.sol

```diff
@@ -0,0 +1,457 @@
+     * turns into:
+     *
+     * bptPrice = from_pool_token(asset_amount).amount * oracle_price / pool_asset_rate
+     *
+     * bptExpected = bptPrice(in relation to specified asset) * asset_amount / BPT_token_rate
+     *
+     * and since from_pool_token(asset_amount).amount and pool_asset_rate cancel each-other out
+     * this makes the final equation:
+     *
+     * bptExpected = oracle_price * asset_amount / BPT_token_rate
+     *
+     * more explanation here:
+     * https://www.notion.so/originprotocol/Support-Balancer-OETH-strategy-9becdea132704e588782a919d7d471eb?pvs=4#382834f9815e46a7937f3acca0f637c5
+     */
+    /* solhint-enable max-line-length */
+    function getBPTExpected(address _asset, uint256 _amount)
+        internal
+        view
+        virtual
+        returns (uint256 bptExpected)
+    {
+        address priceProvider = IVault(vaultAddress).priceProvider();
+        uint256 strategyAssetMarketPrice = IOracle(priceProvider).price(_asset);
+        uint256 bptRate = IRateProvider(platformAddress).getRate();
+
+        bptExpected = _amount
+            .mulTruncate(strategyAssetMarketPrice)
+            .divPrecisely(bptRate);
+    }
+
+    function getBPTExpected(address[] memory _assets, uint256[] memory _amounts)
+        internal
+        view
+        virtual
+        returns (uint256 bptExpected)
+    {
+        // Get the oracle from the OETH Vault
+        address priceProvider = IVault(vaultAddress).priceProvider();
+
+        for (uint256 i = 0; i < _assets.length; ++i) {
+            uint256 strategyAssetMarketPrice = IOracle(priceProvider).price(
+                _assets[i]
+            );
+            // convert asset amount to ETH amount
+            bptExpected =
+                bptExpected +
+                _amounts[i].mulTruncate(strategyAssetMarketPrice);
+        }
+
+        uint256 bptRate = IRateProvider(platformAddress).getRate();
+        // Convert ETH amount to BPT amount
+        bptExpected = bptExpected.divPrecisely(bptRate);
+    }
+
+    function _lpDepositAll() internal virtual;
+
+    function _lpWithdraw(uint256 numBPTTokens) internal virtual;
+
+    function _lpWithdrawAll() internal virtual;
+
+    /**
+     * @notice Balancer returns assets and rateProviders for corresponding assets ordered
+     * by numerical order.
+     */
+    function getPoolAssets() internal view returns (IERC20[] memory assets) {
+        (assets, , ) = balancerVault.getPoolTokens(balancerPoolId);
+    }
+
+    /**
+     * @dev If an asset is rebasing the Balancer pools have a wrapped versions of assets
+     * that the strategy supports. This function converts the pool(wrapped) asset
+     * and corresponding amount to strategy asset.
+     */
+    function toPoolAsset(address asset, uint256 amount)
+        internal
+        view
+        returns (address poolAsset, uint256 poolAmount)
+    {
+        poolAmount = 0;
```

Comment:
Options: Is this line needed? Seems like we don't worry about initializing return variables elsewhere in the code.

Reply:
isn't, removed